upb/decode_fast.c

```diff
@@ -0,0 +1,479 @@
+  len = (int8_t)ptr[tagbytes];
+  str = ptr + tagbytes + 1;
+  dst->data = str;
+  dst->size = len;
+  if (UPB_UNLIKELY(fastdecode_boundscheck(str, len, d->limit))) {
+    dst->size = 0;
+    RETURN_GENERIC("string field len >1 byte\n");
+  }
+  return fastdecode_dispatch(d, str + len, msg, table, hasbits);
+}
+
+const char *upb_pss_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_s);
+}
+
+const char *upb_pos_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_o);
+}
+
+const char *upb_pss_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_s);
+}
+
+const char *upb_pos_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_o);
+}
+
+/* message fields *************************************************************/
+
+UPB_FORCEINLINE
+bool fastdecode_boundscheck2(const char *ptr, unsigned len, const char *end) {
+  uintptr_t uptr = (uintptr_t)ptr;
+  uintptr_t uend = (uintptr_t)end;
+  uintptr_t res = uptr + len;
+  return res < uptr || res > uend;
+}
+
+UPB_FORCEINLINE
+static const char *fastdecode_submsg(UPB_PARSE_PARAMS, int tagbytes,
+                                     int msg_ceil_bytes, upb_card card) {
+  const char *saved_limit = d->limit;
+  const char *saved_fastlimit = d->fastlimit;
+  const upb_msglayout_field *field = &table->fields[data >> 48];
+  size_t ofs = field->offset;
+  const upb_msglayout *subl = table->submsgs[field->submsg_index];
+  upb_array *arr;
+  upb_msg **submsg;
+  void *end;
+
+  if (UPB_UNLIKELY(!fastdecode_checktag(data, tagbytes))) {
+    RETURN_GENERIC("submessage field tag mismatch\n");
+  }
+
+  if (--d->depth < 0) return fastdecode_err(d);
+
+  submsg = fastdecode_getfield_ofs(d, ptr, msg, ofs, &data, &hasbits, &arr,
+                                   &end, tagbytes, sizeof(upb_msg *), card);
+
+again:
+  if (card == CARD_r) {
+    if (UPB_UNLIKELY(submsg == end)) {
+      if (arr) {
+        size_t old_size = arr->size;
+        size_t old_bytes = old_size * sizeof(upb_msg*);
+        size_t new_size = old_size * 2;
+        size_t new_bytes = new_size * sizeof(upb_msg*);
+        char *old_ptr = _upb_array_ptr(arr);
+        if (UPB_UNLIKELY((size_t)(d->arena_end - d->arena_ptr) < new_bytes)) {
+          d->limit = saved_limit;
+          d->fastlimit = saved_fastlimit;
+          arr->len = submsg - (upb_msg**)_upb_array_ptr(arr);
+          d->depth++;
+          RETURN_GENERIC("repeated realloc failed: arena full");
+        }
+        memcpy(d->arena_ptr, old_ptr, old_bytes);
+        arr->size = new_size;
+        arr->data = _upb_array_tagptr(d->arena_ptr, 3);
+        submsg = (void*)(d->arena_ptr + (old_size * sizeof(upb_msg*)));
+        end = (void*)(d->arena_ptr + (new_size * sizeof(upb_msg*)));
+        d->arena_ptr += new_bytes;
+      } else {
+        d->limit = saved_limit;
+        d->fastlimit = saved_fastlimit;
+        d->depth++;
```

Comment:
It seems this path forgets updating arr->len. This should really be handled by goto's, to a single handler that tail calls to generic parser.

Reply:
In this path arr == NULL, so we can't write arr->len.